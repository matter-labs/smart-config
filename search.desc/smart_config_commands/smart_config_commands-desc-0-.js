searchState.loadedDescShard("smart_config_commands", 0, "Command-line extensions for <code>smart-config</code> library.\nReference to a parameter on a configuration.\nWrapper around an I/O writer. Will style the output with …\nIterates over all paths to the param.\nReturns canonical path to the param.\nReference to the configuration containing the param.\nCreates a custom printer.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nParam metadata.\nPrints debug info for all param values in the provided <code>repo</code>…\nPrints help on config params in the provided <code>schema</code>. …\nOutputs JSON with syntax highlighting.\nOutputs YAML adhering to the JSON model with syntax …\nCreates a printer to stderr. The stderr is locked while …\nCreates a printer to stdout. The stdout is locked while …")