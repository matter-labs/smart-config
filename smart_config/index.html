<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="`smart-config` – schema-driven layered configuration system with support of multiple configuration formats."><title>smart_config - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-970bf1bf55bd1c15.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="smart_config" data-themes="" data-resource-suffix="" data-rustdoc-version="1.82.0-nightly (28a58f2fa 2024-07-31)" data-channel="nightly" data-search-js="search-ef54e092fea9c776.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-f9142fe437216fc3.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-df360f571f6edeae.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../smart_config/index.html">smart_<wbr>config</a><span class="version">0.3.0-pre</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Aliases</a></li><li><a href="#derives">Derive Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">smart_config</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/smart_config/lib.rs.html#1-458">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p><code>smart-config</code> – schema-driven layered configuration system with support of multiple configuration formats.</p>
<h2 id="overview"><a class="doc-anchor" href="#overview">§</a>Overview</h2>
<p>The task solved by the library is merging configuration input from a variety of prioritized <a href="trait.ConfigSource.html" title="trait smart_config::ConfigSource">sources</a>
(JSON and YAML files, env variables, command-line args etc.) and converting this input to strongly typed
representation (i.e., config structs or enums). As with other config systems, config input follows the JSON object model
(see <a href="value/enum.Value.html" title="enum smart_config::value::Value"><code>Value</code></a>), with each value enriched with its <a href="value/enum.ValueOrigin.html" title="enum smart_config::value::ValueOrigin">origin</a> (e.g., a path in a specific JSON file,
or a specific env var). This allows attributing errors during deserialization.</p>
<p>The defining feature of <code>smart-config</code> is its schema-driven design. Each config type has associated <a href="metadata/struct.ConfigMetadata.html" title="struct smart_config::metadata::ConfigMetadata">metadata</a>
defined with the help of the <a href="trait.DescribeConfig.html" title="trait smart_config::DescribeConfig"><code>DescribeConfig</code></a> derive macro; deserialization is handled by the accompanying <a href="de/trait.DeserializeConfig.html" title="trait smart_config::de::DeserializeConfig"><code>DeserializeConfig</code></a> macro.
Metadata includes a variety of info extracted from the config type:</p>
<ul>
<li><a href="metadata/struct.ParamMetadata.html" title="struct smart_config::metadata::ParamMetadata">Parameter info</a>: name (including aliases and renaming), help (extracted from doc comments),
type, <a href="de/trait.DeserializeParam.html" title="trait smart_config::de::DeserializeParam">deserializer for the param</a> etc.</li>
<li><a href="metadata/struct.NestedConfigMetadata.html" title="struct smart_config::metadata::NestedConfigMetadata">Nested configurations</a>.</li>
</ul>
<p>Multiple configurations are collected into a global <a href="struct.ConfigSchema.html" title="struct smart_config::ConfigSchema"><code>ConfigSchema</code></a>. Each configuration is <em>mounted</em> at a specific path.
E.g., if a large app has an HTTP server component, it may be mounted at <code>api.http</code>. Multiple config types may be mounted
at the same path (e.g., flattened configs); conversely, a single config type may be mounted at multiple places.
As a result, there doesn’t need to be a god object uniting all configs in the app; they may be dynamically collected and deserialized
inside relevant components.</p>
<p>This information provides rich human-readable info about configs. It also assists when preprocessing and merging config inputs.
For example, env vars are a flat string -&gt; string map; with the help of a schema, it’s possible to:</p>
<ul>
<li>Correctly nest vars (e.g., transform the <code>API_HTTP_PORT</code> var into a <code>port</code> var inside <code>http</code> object inside <code>api</code> object)</li>
<li>Transform value types from strings to expected types.</li>
</ul>
<p>Preprocessing and merging config sources is encapsulated in <a href="struct.ConfigRepository.html" title="struct smart_config::ConfigRepository"><code>ConfigRepository</code></a>.</p>
<h2 id="tldr"><a class="doc-anchor" href="#tldr">§</a>TL;DR</h2>
<ul>
<li>Rich, self-documenting configuration schema.</li>
<li>Utilizes the schema to enrich configuration sources and intelligently merge them.</li>
<li>Doesn’t require a god object uniting all configs in the app; they may be dynamically collected and deserialized
inside relevant components.</li>
<li>Supports lazy parsing for complex / multi-component apps (only the used configs are parsed; other configs are not required).</li>
<li>Supports multiple configuration formats and programmable source priorities (e.g., <code>base.yml</code> + overrides from the
<code>overrides/</code> dir in the alphabetic order + env vars).</li>
<li>Rich and complete deserialization errors including locations and value origins.</li>
<li><a href="de/index.html#secrets" title="mod smart_config::de">Built-in support for secret params</a>.</li>
</ul>
<h2 id="crate-features"><a class="doc-anchor" href="#crate-features">§</a>Crate features</h2><h3 id="primitive-types"><a class="doc-anchor" href="#primitive-types">§</a><code>primitive-types</code></h3>
<p><em>(Off by default)</em></p>
<p>Implements deserialization for basic Ethereum types like <a href="primitive_types::H256"><code>H256</code></a> (32-byte hash)
and <a href="primitive_types::U256"><code>U256</code></a> (256-bit unsigned integer).</p>
<h2 id="examples"><a class="doc-anchor" href="#examples">§</a>Examples</h2><h3 id="basic-workflow"><a class="doc-anchor" href="#basic-workflow">§</a>Basic workflow</h3>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>smart_config::{
    config, ConfigSchema, ConfigRepository, DescribeConfig, DeserializeConfig, Yaml, Environment,
};

<span class="attr">#[derive(Debug, DescribeConfig, DeserializeConfig)]
</span><span class="kw">pub struct </span>TestConfig {
    <span class="kw">pub </span>port: u16,
    <span class="attr">#[config(default_t = <span class="string">"test"</span>.into())]
    </span><span class="kw">pub </span>name: String,
    <span class="attr">#[config(default_t = <span class="bool-val">true</span>)]
    </span><span class="kw">pub </span>tracing: bool,
}

<span class="kw">let </span>schema = ConfigSchema::new(<span class="kw-2">&amp;</span>TestConfig::DESCRIPTION, <span class="string">"test"</span>);
<span class="comment">// Assume we use two config sources: a YAML file and env vars,
// the latter having higher priority.
</span><span class="kw">let </span>yaml = <span class="string">r"
test:
  port: 4000
  name: app
"</span>;
<span class="kw">let </span>yaml = Yaml::new(<span class="string">"test.yml"</span>, serde_yaml::from_str(yaml)<span class="question-mark">?</span>)<span class="question-mark">?</span>;
<span class="kw">let </span>env = Environment::from_iter(<span class="string">"APP_"</span>, [(<span class="string">"APP_TEST_PORT"</span>, <span class="string">"8000"</span>)]);
<span class="comment">// Add both sources to a repo.
</span><span class="kw">let </span>repo = ConfigRepository::new(<span class="kw-2">&amp;</span>schema).with(yaml).with(env);
<span class="comment">// Get the parser for the config.
</span><span class="kw">let </span>parser = repo.single::&lt;TestConfig&gt;()<span class="question-mark">?</span>;
<span class="kw">let </span>config = parser.parse()<span class="question-mark">?</span>;
<span class="macro">assert_eq!</span>(config.port, <span class="number">8_000</span>); <span class="comment">// from the env var
</span><span class="macro">assert_eq!</span>(config.name, <span class="string">"app"</span>); <span class="comment">// from YAML
</span><span class="macro">assert!</span>(config.tracing); <span class="comment">// from the default value</span></code></pre></div>
<h3 id="declaring-type-as-well-known"><a class="doc-anchor" href="#declaring-type-as-well-known">§</a>Declaring type as well-known</h3>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>std::collections::HashMap;
<span class="kw">use </span>smart_config::{
    de::{Serde, WellKnown, WellKnownOption}, metadata::BasicTypes,
    DescribeConfig, DeserializeConfig,
};

<span class="attr">#[derive(Debug, serde::Serialize, serde::Deserialize)]
</span><span class="kw">enum </span>CustomEnum {
    First,
    Second,
}

<span class="kw">impl </span>WellKnown <span class="kw">for </span>CustomEnum {
    <span class="comment">// signals that the type should be deserialized via `serde`
    // and the expected input is a string
    </span><span class="kw">type </span>Deserializer = <span class="macro">Serde!</span>[str];
    <span class="kw">const </span>DE: <span class="self">Self</span>::Deserializer = <span class="macro">Serde!</span>[str];
}

<span class="comment">// Signals that the type can be used with an `Option&lt;_&gt;`
</span><span class="kw">impl </span>WellKnownOption <span class="kw">for </span>CustomEnum {}

<span class="comment">// Then, the type can be used in configs basically everywhere:
</span><span class="attr">#[derive(Debug, DescribeConfig, DeserializeConfig)]
</span><span class="kw">struct </span>TestConfig {
    value: CustomEnum,
    optional: <span class="prelude-ty">Option</span>&lt;CustomEnum&gt;,
    repeated: Vec&lt;CustomEnum&gt;,
    map: HashMap&lt;String, CustomEnum&gt;,
}</code></pre></div>
</div></details><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name" id="reexport.DeserializeConfig"><code>pub use self::de::<a class="trait" href="de/trait.DeserializeConfig.html" title="trait smart_config::de::DeserializeConfig">DeserializeConfig</a>;</code></div></li><li><div class="item-name" id="reexport.ParseErrorCategory"><code>pub use self::error::ParseErrorCategory;</code></div></li></ul><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="de/index.html" title="mod smart_config::de">de</a></div><div class="desc docblock-short">Configuration deserialization logic.</div></li><li><div class="item-name"><a class="mod" href="fallback/index.html" title="mod smart_config::fallback">fallback</a></div><div class="desc docblock-short">Fallback <a href="value/enum.Value.html" title="enum smart_config::value::Value"><code>Value</code></a> sources.</div></li><li><div class="item-name"><a class="mod" href="metadata/index.html" title="mod smart_config::metadata">metadata</a></div><div class="desc docblock-short">Configuration metadata.</div></li><li><div class="item-name"><a class="mod" href="testing/index.html" title="mod smart_config::testing">testing</a></div><div class="desc docblock-short">Testing tools for configurations.</div></li><li><div class="item-name"><a class="mod" href="validation/index.html" title="mod smart_config::validation">validation</a></div><div class="desc docblock-short">Parameter and config validation and filtering.</div></li><li><div class="item-name"><a class="mod" href="value/index.html" title="mod smart_config::value">value</a></div><div class="desc docblock-short">Enriched JSON object model that allows to associate values with origins.</div></li><li><div class="item-name"><a class="mod" href="visit/index.html" title="mod smart_config::visit">visit</a></div><div class="desc docblock-short">Visitor pattern for configs.</div></li></ul><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.Serde.html" title="macro smart_config::Serde">Serde</a></div><div class="desc docblock-short">Constructor of <a href="de/struct.Serde.html" title="struct smart_config::de::Serde"><code>Serde</code></a> types / instances.</div></li><li><div class="item-name"><a class="macro" href="macro.config.html" title="macro smart_config::config">config</a></div><div class="desc docblock-short">Creates <a href="struct.Json.html" title="struct smart_config::Json"><code>Json</code></a> configuration input based on the provided list of path–value tuples.
This is essentially a slightly fancier / more specialized version of <a href="https://docs.rs/serde_json/1.0.133/serde_json/macro.json.html" title="macro serde_json::json"><code>json!</code></a>.</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ByteSize.html" title="struct smart_config::ByteSize">Byte<wbr>Size</a></div><div class="desc docblock-short">A wrapper providing a clear reminder that the wrapped value represents the number of bytes.</div></li><li><div class="item-name"><a class="struct" href="struct.ConfigMut.html" title="struct smart_config::ConfigMut">Config<wbr>Mut</a></div><div class="desc docblock-short">Mutable reference to a specific configuration inside <a href="struct.ConfigSchema.html" title="struct smart_config::ConfigSchema"><code>ConfigSchema</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.ConfigParser.html" title="struct smart_config::ConfigParser">Config<wbr>Parser</a></div><div class="desc docblock-short">Parser of configuration input in a <a href="struct.ConfigRepository.html" title="struct smart_config::ConfigRepository"><code>ConfigRepository</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.ConfigRef.html" title="struct smart_config::ConfigRef">Config<wbr>Ref</a></div><div class="desc docblock-short">Reference to a specific configuration inside <a href="struct.ConfigSchema.html" title="struct smart_config::ConfigSchema"><code>ConfigSchema</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.ConfigRepository.html" title="struct smart_config::ConfigRepository">Config<wbr>Repository</a></div><div class="desc docblock-short">Configuration repository containing zero or more <a href="trait.ConfigSource.html" title="trait smart_config::ConfigSource">configuration sources</a>.
Sources are preprocessed and merged according to the provided <a href="struct.ConfigSchema.html" title="struct smart_config::ConfigSchema"><code>ConfigSchema</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.ConfigSchema.html" title="struct smart_config::ConfigSchema">Config<wbr>Schema</a></div><div class="desc docblock-short">Schema for configuration. Can contain multiple configs bound to different paths.</div></li><li><div class="item-name"><a class="struct" href="struct.ConfigSources.html" title="struct smart_config::ConfigSources">Config<wbr>Sources</a></div><div class="desc docblock-short">Prioritized list of configuration sources. Can be used to push multiple sources at once
into a <a href="struct.ConfigRepository.html" title="struct smart_config::ConfigRepository"><code>ConfigRepository</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.DeserializeConfigError.html" title="struct smart_config::DeserializeConfigError">Deserialize<wbr>Config<wbr>Error</a></div><div class="desc docblock-short">Marker error for <a href="de/trait.DeserializeConfig.html" title="trait smart_config::de::DeserializeConfig"><code>DeserializeConfig</code></a> operations. The error info os stored
in <a href="de/struct.DeserializeContext.html" title="struct smart_config::de::DeserializeContext"><code>DeserializeContext</code></a> as <a href="struct.ParseErrors.html" title="struct smart_config::ParseErrors"><code>ParseErrors</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.Environment.html" title="struct smart_config::Environment">Environment</a></div><div class="desc docblock-short">Configuration sourced from environment variables.</div></li><li><div class="item-name"><a class="struct" href="struct.Flat.html" title="struct smart_config::Flat">Flat</a></div><div class="desc docblock-short">Marker for key–value / flat configuration sources (e.g., env variables or command-line args).</div></li><li><div class="item-name"><a class="struct" href="struct.Hierarchical.html" title="struct smart_config::Hierarchical">Hierarchical</a></div><div class="desc docblock-short">Marker for hierarchical configuration sources (e.g. JSON or YAML files).</div></li><li><div class="item-name"><a class="struct" href="struct.Json.html" title="struct smart_config::Json">Json</a></div><div class="desc docblock-short">JSON-based configuration source.</div></li><li><div class="item-name"><a class="struct" href="struct.ParseError.html" title="struct smart_config::ParseError">Parse<wbr>Error</a></div><div class="desc docblock-short">Config parameter deserialization errors.</div></li><li><div class="item-name"><a class="struct" href="struct.ParseErrors.html" title="struct smart_config::ParseErrors">Parse<wbr>Errors</a></div><div class="desc docblock-short">Collection of <a href="struct.ParseError.html" title="struct smart_config::ParseError"><code>ParseError</code></a>s returned from <a href="struct.ConfigParser.html#method.parse" title="method smart_config::ConfigParser::parse"><code>ConfigParser::parse()</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.Prefixed.html" title="struct smart_config::Prefixed">Prefixed</a></div><div class="desc docblock-short">Wraps a hierarchical source into a prefix.</div></li><li><div class="item-name"><a class="struct" href="struct.SerializerOptions.html" title="struct smart_config::SerializerOptions">Serializer<wbr>Options</a></div><div class="desc docblock-short">Configuration serialization options.</div></li><li><div class="item-name"><a class="struct" href="struct.SourceInfo.html" title="struct smart_config::SourceInfo">Source<wbr>Info</a></div><div class="desc docblock-short">Information about a source returned from <a href="struct.ConfigRepository.html#method.sources" title="method smart_config::ConfigRepository::sources"><code>ConfigRepository::sources()</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.Yaml.html" title="struct smart_config::Yaml">Yaml</a></div><div class="desc docblock-short">YAML-based configuration source.</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.ConfigSource.html" title="trait smart_config::ConfigSource">Config<wbr>Source</a></div><div class="desc docblock-short">Source of configuration parameters that can be added to a <a href="struct.ConfigRepository.html" title="struct smart_config::ConfigRepository"><code>ConfigRepository</code></a>.</div></li><li><div class="item-name"><a class="trait" href="trait.ConfigSourceKind.html" title="trait smart_config::ConfigSourceKind">Config<wbr>Source<wbr>Kind</a></div><div class="desc docblock-short">Kind of a <a href="trait.ConfigSource.html" title="trait smart_config::ConfigSource"><code>ConfigSource</code></a>.</div></li><li><div class="item-name"><a class="trait" href="trait.DescribeConfig.html" title="trait smart_config::DescribeConfig">Describe<wbr>Config</a></div><div class="desc docblock-short">Describes a configuration (i.e., a group of related parameters).</div></li><li><div class="item-name"><a class="trait" href="trait.ExampleConfig.html" title="trait smart_config::ExampleConfig">Example<wbr>Config</a></div><div class="desc docblock-short">Provides an example for this configuration. The produced config can be used in tests etc.</div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.ErrorWithOrigin.html" title="type smart_config::ErrorWithOrigin">Error<wbr>With<wbr>Origin</a></div><div class="desc docblock-short">Error together with its origin.</div></li></ul><h2 id="derives" class="section-header">Derive Macros<a href="#derives" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="derive" href="derive.DescribeConfig.html" title="derive smart_config::DescribeConfig">Describe<wbr>Config</a></div><div class="desc docblock-short">Derives the <a href="trait.DescribeConfig.html" title="trait smart_config::DescribeConfig"><code>DescribeConfig</code></a> trait for a type.</div></li><li><div class="item-name"><a class="derive" href="derive.DeserializeConfig.html" title="derive smart_config::DeserializeConfig">Deserialize<wbr>Config</a></div><div class="desc docblock-short">Derives the <a href="de/trait.DeserializeConfig.html" title="trait smart_config::de::DeserializeConfig"><code>DeserializeConfig</code></a> trait for a type.</div></li><li><div class="item-name"><a class="derive" href="derive.ExampleConfig.html" title="derive smart_config::ExampleConfig">Example<wbr>Config</a></div><div class="desc docblock-short">Derives the <a href="trait.ExampleConfig.html" title="trait smart_config::ExampleConfig"><code>ExampleConfig</code></a> trait for a type.</div></li></ul></section></div></main></body></html>