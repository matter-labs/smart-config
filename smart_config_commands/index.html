<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Command-line extensions for `smart-config` library."><title>smart_config_commands - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-970bf1bf55bd1c15.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="smart_config_commands" data-themes="" data-resource-suffix="" data-rustdoc-version="1.82.0-nightly (28a58f2fa 2024-07-31)" data-channel="nightly" data-search-js="search-ef54e092fea9c776.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-f9142fe437216fc3.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-df360f571f6edeae.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../smart_config_commands/index.html">smart_<wbr>config_<wbr>commands</a><span class="version">0.3.0-pre</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#structs">Structs</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">smart_config_commands</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/smart_config_commands/lib.rs.html#1-149">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Command-line extensions for <code>smart-config</code> library.</p>
<p>The extensions are as follows:</p>
<ul>
<li><a href="struct.Printer.html#method.print_help" title="method smart_config_commands::Printer::print_help">Printing help</a> for configuration params with optional filtering.</li>
<li><a href="struct.Printer.html#method.print_debug" title="method smart_config_commands::Printer::print_debug">Debugging</a> param values and deserialization errors.</li>
</ul>
<p>All extensions are encapsulated in <a href="struct.Printer.html" title="struct smart_config_commands::Printer"><code>Printer</code></a>.</p>
<h2 id="examples"><a class="doc-anchor" href="#examples">§</a>Examples</h2>
<p>See the crate readme or the <code>examples</code> dir for captured output samples.</p>
<h3 id="printing-help"><a class="doc-anchor" href="#printing-help">§</a>Printing help</h3>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>smart_config::ConfigSchema;
<span class="kw">use </span>smart_config_commands::Printer;

<span class="kw">let </span><span class="kw-2">mut </span>schema = ConfigSchema::default();
<span class="comment">// Add configurations to the schema...

</span>Printer::stderr().print_help(<span class="kw-2">&amp;</span>schema, |<span class="kw">_</span>| <span class="bool-val">true</span>)<span class="question-mark">?</span>;</code></pre></div>
<h3 id="debugging-param-values"><a class="doc-anchor" href="#debugging-param-values">§</a>Debugging param values</h3>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>smart_config::{ConfigSchema, ConfigRepository};
<span class="kw">use </span>smart_config_commands::Printer;

<span class="kw">let </span><span class="kw-2">mut </span>schema = ConfigSchema::default();
<span class="comment">// Add configurations to the schema...
</span><span class="kw">let </span><span class="kw-2">mut </span>repo = ConfigRepository::new(<span class="kw-2">&amp;</span>schema);
<span class="comment">// Add sources to the repository...

</span>Printer::stderr().print_debug(<span class="kw-2">&amp;</span>repo, |param_ref| {
    <span class="comment">// Allows filtering output params
    </span>param_ref.canonical_path().starts_with(<span class="string">"test."</span>)
})<span class="question-mark">?</span>;</code></pre></div>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ParamRef.html" title="struct smart_config_commands::ParamRef">Param<wbr>Ref</a></div><div class="desc docblock-short">Reference to a parameter on a configuration.</div></li><li><div class="item-name"><a class="struct" href="struct.Printer.html" title="struct smart_config_commands::Printer">Printer</a></div><div class="desc docblock-short">Wrapper around an I/O writer. Will style the output with ANSI sequences if appropriate.</div></li></ul></section></div></main></body></html>