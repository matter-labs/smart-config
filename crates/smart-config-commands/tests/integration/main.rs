//! **Important.** The generated snapshots are specific to stable Rust; nightly Rust provides better spanning for types
//! (`Option<u64>` instead of `Option`).

use std::fmt;

use anstream::AutoStream;
use smart_config::{ConfigSchema, DescribeConfig, Environment, ExampleConfig, SerializerOptions};
use smart_config_commands::Printer;
use test_casing::{test_casing, Product};

use crate::configs::{create_mock_repo, TestConfig};

mod configs;

#[test]
fn full_config_help() {
    let schema = ConfigSchema::new(&TestConfig::DESCRIPTION, "test");

    let mut buffer = vec![];
    Printer::custom(AutoStream::never(&mut buffer))
        .print_help(&schema, |_| true)
        .unwrap();
    let buffer = String::from_utf8(buffer).unwrap();
    insta::assert_snapshot!("help_full", buffer);
}

#[test]
fn filtered_config_help() {
    let schema = ConfigSchema::new(&TestConfig::DESCRIPTION, "test");

    let mut buffer = vec![];
    Printer::custom(AutoStream::never(&mut buffer))
        .print_help(&schema, |param| {
            param.all_paths().any(|(path, _)| path.contains("fund"))
        })
        .unwrap();
    let buffer = String::from_utf8(buffer).unwrap();
    insta::assert_snapshot!("help_filtered", buffer);
}

#[test]
fn full_config_debug() {
    let schema = ConfigSchema::new(&TestConfig::DESCRIPTION, "test");
    let repo = create_mock_repo(&schema, false);

    let mut buffer = vec![];
    Printer::custom(AutoStream::never(&mut buffer))
        .print_debug(&repo, |_| true)
        .unwrap()
        .unwrap();
    let buffer = String::from_utf8(buffer).unwrap();
    insta::assert_snapshot!("debug_full", buffer);
}

#[test]
fn erroneous_config_debug() {
    let schema = ConfigSchema::new(&TestConfig::DESCRIPTION, "test");
    let repo = create_mock_repo(&schema, true);

    let mut buffer = vec![];
    Printer::custom(AutoStream::never(&mut buffer))
        .print_debug(&repo, |_| true)
        .unwrap()
        .unwrap_err();
    let buffer = String::from_utf8(buffer).unwrap();
    insta::assert_snapshot!("debug_errors", buffer);
}

#[derive(Debug, Clone, Copy)]
enum Format {
    Yaml,
    Json,
    Env,
}

impl fmt::Display for Format {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(match self {
            Self::Yaml => "yaml",
            Self::Json => "json",
            Self::Env => "env",
        })
    }
}

impl Format {
    const ALL: [Self; 3] = [Self::Yaml, Self::Json, Self::Env];
}

#[test_casing(6, Product((Format::ALL, [false, true])))]
#[test]
fn printing_config_example_yaml(format: Format, diff: bool) {
    let example_config = TestConfig::example_config();
    let options = if diff {
        SerializerOptions::diff_with_default()
    } else {
        SerializerOptions::default()
    };
    let options = options.flat(matches!(format, Format::Env));
    let example_json = options.serialize(&example_config);

    let mut buffer = vec![];
    match format {
        Format::Yaml => {
            Printer::custom(AutoStream::never(&mut buffer))
                .print_yaml(&example_json.into())
                .unwrap();
        }
        Format::Json => {
            Printer::custom(AutoStream::never(&mut buffer))
                .print_json(&example_json.into())
                .unwrap();
        }
        Format::Env => {
            let env = Environment::convert_flat_params(&example_json, "APP_");
            Printer::custom(AutoStream::never(&mut buffer))
                .print_yaml(&env.into())
                .unwrap();
        }
    }

    let buffer = String::from_utf8(buffer).unwrap();
    let snapshot_name = format!("example_{}_{format}", if diff { "diff" } else { "full" });
    insta::assert_snapshot!(snapshot_name, buffer);
}
