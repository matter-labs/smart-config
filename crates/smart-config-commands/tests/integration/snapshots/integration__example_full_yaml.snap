---
source: crates/smart-config-commands/tests/integration/main.rs
assertion_line: 124
expression: buffer
---
app_name: app
cache_size: 16 MiB
dir_paths:
  - ./local
funding:
  address: '0x0000000000000000000000000000000000000000'
  api_key: correct horse battery staple
  balance: '0x0'
  secret_key: '0x0000000000000000000000000000000000000000000000000000000000000000'
nested:
  complex:
    array:
      - 3
      - 5
    map:
      var: 3
  exit_on_error: true
  method_limits:
    - method: eth_blockNumber
      rps: 1
    - method: eth_call
      rps: 100
  more_timeouts:
    - 5s
object_store:
  path: .
  type: local
poll_latency: 500ms
port: 8080
required: 42
scaling_factor: 0.5
temp_dir: /tmp
timeout_sec: 60
